### PR TITLE
std_misc/thread: additional test-case example

### PR DESCRIPTION
**Pull request by juleskers**
As discussed in #859, I've reprocessed my code into a "test-case" for the thread example.

Try the new example on the [alternative playground](http://play.integer32.com/?gist=ccd47b4c20862d3735aa88f6ccacde8d&version=undefined)

I haven't tried to build this yet, as I'm missing a lot of infrastructure on this current PC.

### CODE REVIEW

**Comment by steveklabnik** [comment]
Hey @juleskers ! Just so you know, it seems like you haven't set up git/github properly; all those commits don't have your avatar. Is that intentional?

**Comment by juleskers** [comment]
@steveklabnik oh ups, that's definitely not intentional; that's my work git-user, not my personal git... Still me though.
I'll make a rebase with the my personal/github-linked commits in a new PR.

As for the Travis failure: seems legit, though I can't resist making puns on "double standards":
> ./check-line-length.sh
Checking if any rust file has a line longer than _**99**_ characters
  \> ./examples/std_misc/threads/testcase_mapreduce/mapreduce.rs exceeds _**79**_ chars

(Emphasis mine)

All documentation I encountered, including `contributing.md` mention 99-char line-lengths, but it seems Travis fails on 80. Anyone know why?

**Comment by juleskers** [comment]
OK, so I lied and lived dangerously with `push --force` :smiling_imp: 
My apologies to Travis for providing double-work

**Comment by juleskers** [comment]
OK, reducing all line-widths to 80 seems to fix things.

- [contributing.md](https://github.com/rust-lang/rust-by-example/blob/master/CONTRIBUTING.md#rust-code-rs) says rust-code can have 99 chars.
- the check-line-lenght script [checks `suspects`](https://github.com/rust-lang/rust-by-example/blob/master/check-line-length.sh#L10) with lines longer than 100 chars, but [prints](https://github.com/rust-lang/rust-by-example/blob/master/check-line-length.sh#L20:L21) every line longer than 79, but _only_ if at least one line is >=100.. This should probably be fixed.

File: examples/std_misc/threads/testcase_mapreduce/mapreduce.rs [hunk 1]
```diff
@@ -0,0 +1,95 @@
+use std::thread;
+
+// This is the `main` thread
+fn main() {
+
+    // This is our data to process.
+    // We will calculate the sum of all digits via a threaded 
+    // map-reduce algorithm.
+    // each whitespace separated chunk will be handled in a different thread.
+    // TODO: see what happens to the output if you insert spaces!
```
Comment (mdinger) [inline examples/std_misc/threads/testcase_mapreduce/mapreduce.rs]: I would recommend you attempt to keep the comment right edge close to aligned. The same for the left. Introduction of new paragraphs could be done with a blank comment line.

**Comment by mdinger** [comment]
If you're interested in further extensions, you could add an *activities* section which [some pages](http://rustbyexample.com/hello/print.html) contain. This could be a recommendation of interesting extensions people could attempt to do with it. I definitely would say it is required though.

**Comment by juleskers** [comment]
Done and Done!

As for spelling of "[parallelise](http://www.dictionary.com/browse/parallelise)", I double-checked. 's' or 'z' both work, depending if we want British or American English.
( "paralyze" means "to render motionless", which I'm betting we don't need ;-) )

File: examples/std_misc/threads/testcase_mapreduce/input.md [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+Note that, although we're passing references across thread boundaries, Rust understands that we're
+only passing read-only references, and that thus no unsafety or data races can occur. Because
+we're `move`-ing the data segments into the thread, Rust will also ensure the data is kept alive
+until the threads exit, so no dangling pointers occur.
+
+{mapreduce.play}
+
+### Assignments
+It is not wise to let our number of threads depend on user inputted data.
+What if the user decides to insert a lot of spaces? Do we _really_ want to spawn 2.000 threads?
```
Comment (mdinger) [inline examples/std_misc/threads/testcase_mapreduce/input.md]: 2,000 probably since this is english.

**Comment by mdinger** [comment]
Looks good to me and the assignment looks really good.

@juleskers Thanks for your efforts. @steveklabnik will eventually get around to reviewing for the final say.

**Comment by juleskers** [comment]
Fixed the "2,000". Now we wait for Steve!

**Comment by juleskers** [comment]
Ping @steveklabnik ? :-)

**Comment by steveklabnik** [comment]
Hi! So sorry this took me an absurdly long time. Thank you so much for all the work here.

**Comment by juleskers** [comment]
That's allright! I saw how you are coordinating/communicating the libz blitz AND the new rust book.

Thanks for taking a few moments out of your extremely busy schedule!

**Comment by steveklabnik** [comment]
:heart:

**Pull request merged**